hw/opentitan/ot_usbdev.c

```diff
@@ -664,6 +1011,315 @@ static void ot_usbdev_simulate_link_reset(OtUsbdevState *s)
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: %s: write to CONFIGIN%d with invalid BUFFER (%u)\n",
+                      __func__, s->ot_id, epnum, buf_id);
+        return;
+    }
+
+    trace_ot_usbdev_packet_sent(s->ot_id, epnum, buf_id, pkt_size);
+
+    /* Mark the packet as sent and acknowledged by the host */
+    configin = SHARED_FIELD_DP32(configin, CONFIGIN_RDY, 0u);
+    s->regs[reg] = configin;
+    s->regs[R_IN_DATA_TOGGLE] ^=
+        FIELD_DP32(0u, IN_DATA_TOGGLE, STATUS, 1u << epnum);
+    s->regs[R_IN_SENT] |= 1u << epnum;
+    ot_usbdev_update_irqs(s);
+
+    /* Check for overflow */
+    if (pkt_size > xfer->max_packet_size) {
+        ot_usbdev_complete_transfer_in(
+            s, epnum, OT_USBDEV_SERVER_STATUS_ERROR,
+            "device sent packet bigger than max packet size");
+        return;
+    }
+    if (pkt_size > xfer->recv_rem) {
+        ot_usbdev_complete_transfer_in(s, epnum, OT_USBDEV_SERVER_STATUS_ERROR,
+                                       "device sent more data than expected");
+        return;
+    }
+
+    /* Copy data and advance buffers */
+    ot_usbdev_read_buffer(s, buf_id, xfer->recv_buf, pkt_size);
+
+    xfer->recv_buf += pkt_size;
+    xfer->recv_rem -= pkt_size;
+
+    /* Handle completion */
+    if (pkt_size < xfer->max_packet_size || xfer->recv_rem == 0u) {
+        ot_usbdev_complete_transfer_in(s, epnum,
+                                       OT_USBDEV_SERVER_STATUS_SUCCESS,
+                                       "success");
+        return;
+    }
+}
+
+static void ot_usbdev_advance_transfer_out(OtUsbdevState *s, uint8_t epnum)
+{
+    g_assert(epnum < USBDEV_PARAM_N_ENDPOINTS);
+    /* See BFM (out_packet) */
+
+    /* Nothing to do if there is no pending transfer */
+    OtUsbdevServer *server = &s->usb_server;
+    OtUsbdevServerEpOutXfer *xfer = &server->ep[epnum].out;
+    if (!xfer->pending) {
+        return;
+    }
+
+    /*
+     * If the endpoint OUT is not enabled, the device will silently ignored the
+     * packet, causing an error.
+     */
+    if (!ot_usbdev_is_ep_out_enabled(s, epnum)) {
+        ot_usbdev_complete_transfer_out(s, epnum, OT_USBDEV_SERVER_STATUS_ERROR,
+                                        "endpoint OUT not enabled");
+        return;
+    }
+
+    /* Check for STALL */
+    if (ot_usbdev_is_ep_out_stalled(s, epnum)) {
+        ot_usbdev_complete_transfer_out(s, epnum,
+                                        OT_USBDEV_SERVER_STATUS_STALLED,
+                                        "stalled by device");
+        return;
+    }
+
+    /*
+     * Only accept the DATA packet if the endpoint is enabled, and there is
+     * buffer available and there is space in the RX FIFO. The last entry of the
+     * FIFO is reserved for SETUP packets.
+     */
+    if (!ot_usbdev_is_ep_out_rxenabled(s, epnum) ||
+        fifo8_is_empty(&s->av_out_fifo) || fifo32_num_free(&s->rx_fifo) <= 1) {
+        return;
```

Comment:
Is this not needed due to cancelling an OUT transaction?
```suggestion
        /* some sort of trace would be helpful? */ 
        s->regs[R_USBDEV_INTR_STATE] |= USBDEV_INTR_LINK_OUT_ERR_MASK;
        ot_usbdev_update_irqs(s);
        return;
```
And does this need a way to give a NAK response, is this still todo/WIP or not going to be emulated? (or maybe I don't understand USB)

Reply:
I am not sure what your question is exactly: the way my driver works, a NAK means doing nothing (transfer does not progress but is not cancelled) and it will be attempted again the next time the situation changes. Here for example if RX is disabled, the function returns early and does not nothing. When the driver writes the register to enable RX, this function will be called again to see if the transfer can progress.

Also SETUP packets are special: a device cannot NAK or STALL a SETUP, only silently drop it. This is why the protocol does not have a command to answer a `Setup`.

Reply:
I see, that makes a lot more sense, thanks for explaining.

(I think that is separate to the link out error which is still needed though?)

Reply:
Yes the link error is needed: even the host has no way of knowing that the SETUP packet was dropped, the device does know (even though the interrupt is quite general and the IP does not track the source of errors).

Reply:
I think you still did not add this link error? (maybe I am missing it somehow)

Reply:
Sorry my comment was not very clear. I believe the code matches the RTL here, I will add some comments to clarify it.

Reply:
I am still somewhat confused by this, to be honest. Comparing to the BFM in the `out_packet` function `PidTypeOutToken` case if the available out buffer is empty, there is not space in the RX fifo or the endpoint is not RX enabled, then it calls `cancel_out()` which sets the interrupt for a link out error.

If I follow the interrupt signal through the RTL I end up in `hw/ip/usbdev/rtl/usb_fs_nb_out_pe.sv` where we set `rollback_data` (which triggers the interrupt) to 1 on any of:
 (a) a bad data toggle (while also not stalled).
 (b) Malformed (invalid / non-data) packets received.
 (c) Stalled EP (for a non-isochronous OUT transaction).
 (d) Transactions that are NAK'd due to the EP being full (for either SETUP, OUT or isochronous transactions, just at different stages). But it looks like the EP is only reported as full if either (i) *all* endpoints are blocked because the HW cannot take transactions at all (AV is empty, RX is full) or (ii) the endpoint is blocked (RX out disabled, or not yet setup).

So based on that, I think the BFM is correct here? But also maybe there should be a link out error on the stall case?

Reply:
Indeed you are correct that I misread the BFM and it should trigger the error when it cannot store the packet. It's a bit concerning about STALL however, it seems that it would trigger as link out error on stall which is not documented and really unexpected.

Reply:
Thanks, I think I agree with your current implementation (maybe there should be more link out errors on some error cases e.g. malformed packets but I'm not 100% sure for now).
For reference, here's a link to the RTL line that should trigger a Link Out Error if there is a stall for the EP on reception of the OUT Packet: https://github.com/lowRISC/opentitan/blob/2c4c18ba24c13de66da8cebaf3b38add6a1fd83c/hw/ip/usbdev/rtl/usb_fs_nb_out_pe.sv#L291